### PR TITLE
add blabber protocol support

### PR DESCRIPTION
**Pull request by brayniac**
Adds support for the `blabber` protocol in which the server will periodically publish messages to subscribers.

This is implemented as a pubsub protocol.

Adds latency distribution information to the dataspec for pubsub.

### CODE REVIEW

File: src/config/pubsub.rs [hunk 1]
```diff
@@ -14,6 +14,19 @@ pub struct Pubsub {
     publisher_poolsize: usize,
     publisher_concurrency: usize,
 
+    /// Specify the default sizes for the read and write buffers (in bytes).
+    /// It is useful to increase the sizes if you expect to send and/or receive
+    /// large requests/responses as part of the workload.
+    ///
+    /// The default is a single page (4KB) for each the read and write buffers.
```
Comment (mihirn) [inline src/config/pubsub.rs:21]: Isn't the default buffer size 4 pages?

File: src/config/pubsub.rs [hunk 2]
```diff
@@ -48,6 +61,19 @@ impl Pubsub {
         self.publisher_concurrency
     }
 
+    pub fn read_buffer_size(&self) -> usize {
```
Comment (mihirn) [inline src/config/pubsub.rs:64]: Not a blocker, but I wonder if the definitions and these functions should be moved into their own structure which can be shared between `Client` and `Pubsub`.

File: lib/dataspec/src/lib.rs [hunk 3]
```diff
@@ -64,6 +64,10 @@ pub struct ClientStats {
 pub struct PubsubStats {
     pub publishers: Publishers,
     pub subscribers: Subscribers,
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (mihirn) [inline lib/dataspec/src/lib.rs:67]: Do we also want other stats like ok, corrupt, invalid or is just the count and latency sufficient?

File: src/pubsub/blabber.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                    PUBSUB_SUBSCRIBER_CURR.add(1);
+                    PUBSUB_SUBSCRIBE_OK.increment();
+
+                    Some(s)
+                }
+                Ok(Err(_)) => {
+                    CONNECT_EX.increment();
+                    PUBSUB_SUBSCRIBE_EX.increment();
+
+                    sleep(Duration::from_millis(100)).await;
```
Comment (mihirn) [inline src/pubsub/blabber.rs:82]: Should we increase the backoff over time in case the connection keeps failing?

File: src/pubsub/blabber.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                                    let _ = validator.validate(&mut mbuf);
+
+                                    // return the number of bytes consumed
+                                    len
+                                } else {
+                                    break;
+                                }
+                            };
+
+                            read_buffer.advance(consumed);
```
Comment (mihirn) [inline src/pubsub/blabber.rs:129]: Should we have a `PUBSUB_RECEIVE.increment()` and a `PUBSUB_RECEIVE_OK.increment()` here?

File: src/pubsub/blabber.rs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+async fn publisher_task(work_receiver: Receiver<WorkItem>) -> Result<()> {
+    PUBSUB_PUBLISHER_CURR.add(1);
+
+    while RUNNING.load(Ordering::Relaxed) {
+        let _work_item = work_receiver
+            .recv()
+            .await
+            .map_err(|_| Error::new(ErrorKind::Other, "channel closed"))?;
+    }
+
```
Comment (mihirn) [inline src/pubsub/blabber.rs:168]: Don't we actually need to publish anything here? Or where is that actually happening?

File: src/config/pubsub.rs (hunk 1 above)
Comment (brayniac) [reply src/config/pubsub.rs]: Yes. Fixed

File: src/config/pubsub.rs (hunk 2 above)
Comment (brayniac) [reply src/config/pubsub.rs:64]: That's a good idea. Will address in a follow-up PR.

File: src/pubsub/blabber.rs (hunk 4 above)
Comment (brayniac) [reply src/pubsub/blabber.rs:82]: I think we should address this more holistically in a follow-up PR.

File: src/pubsub/blabber.rs (hunk 6 above)
Comment (brayniac) [reply src/pubsub/blabber.rs]: No. The server actively publishes without the client doing anything. I've simplified by removing all publisher tasks and adding a comment that hopefully makes this cleared going-forward.

File: src/pubsub/blabber.rs (hunk 5 above)
Comment (brayniac) [reply src/pubsub/blabber.rs:129]: This happens up one level in the MessageValidator.

Comment (mihirn) [reply src/pubsub/blabber.rs:129]: Ack, thanks.

**Review by mihirn — APPROVED** [review]

File: lib/dataspec/src/lib.rs (hunk 3 above)
Comment (brayniac) [reply lib/dataspec/src/lib.rs:67]: We should add the rest of the publisher and subscriber stats to this. I'll leave it for another PR though.

**Pull request merged**